### ISSUE TITLE
Add TorchDynamo as a submodule to Pytorch?

### ISSUE THREAD

**Issue by kevinstephano** [pytorch/pytorch#80742]
### 🚀 The feature, motivation and pitch

It has been hard to recommend TorchDynamo usage with Pytorch given that TorchDynamo does not have an official release and users often want to use specific versions of Pytorch (Nvidia releases containers on a monthly basis from TOT), however, TorchDynamo requires using Pytorch TOT, currently, and can break when not in sync.  

I wanted to propose adding TorchDynamo as a submodule to Pytorch.

**Comment by kevinstephano** [comment]
CC @jansel @csarofeen

**Comment by jansel** [comment]
I like this idea, cc @ezyang @malfet

**Comment by csarofeen** [comment]
I think we should make sure to target this in time for the next PyTorch release. Would be really helpful to get people using Dynamo!

**Comment by ezyang** [comment]
Note we already have a hash pin for CI at .github/ci_commit_pins/torchdynamo.txt which you can use to get a known working copy of torchdynamo. I suppose we could upgrade it into a submodule but there is not too much difference; the real question is if you want to start packaging/building torchdynamo by default.

**Comment by csarofeen** [comment]
Do you have one for FuncTorch as well?

**Comment by ezyang** [comment]
We could but PoR is functorch is getting put into PyTorch main repo directly @zou3519

**Comment by zou3519** [comment]
Soon we're putting functorch into the PyTorch repo. By 1.13 we aim to package functorch with the PyTorch release binaries. This helps us (1) clean up the user story -- no need to install 2 packages and (2) saves us from doing a separate release, which is resource intensive.

**Comment by csarofeen** [comment]
Is there any reason we wouldn't want to follow suit with Dynamo?

**Comment by ezyang** [comment]
It's mostly to keep dynamo moving faster without having to wait for full PT CI. Probably when it slows down this would make sense

**Comment by ezyang** [comment]
To make it easier for people to use torchdynamo, there are two tiers:

1. People who can build from source
2. People who need nightlies

For a build from source, we already have torchdynamo commit hash, and we could add a build flag to setup.py to checkout and also build torchdynamo. For functorch, we don't currently have a hash but could add one, this will be mooted when functorch gets merged into pytorch/pytorch officially.

For nightlies, its probably easiest if we have a single nightly that contains all the pieces. We probably can setup some special setup to make this happen.

@csarofeen do you think this solution fulfills the requirements? Is (1) or (2) more important?

**Comment by csarofeen** [comment]
I think this would be really helpful and checks all boxes.

Having a nightly that people can easily grab and try out would make iterating with multiple teams really easy.

This would mean the install process would be fairly straight forward (because why not), this means we can iterate on however many forks would suit the many folks working on different areas.

**Comment by kevinstephano** [comment]
I think (1.) supporting build from source might be more important to us if (2.) means making a specific nightly since we build our own specific nightly and monthly builds.

**Comment by ezyang** [comment]
OK. I don't know how you build your nightlies, but would a series of commands that also gets you a known working torchdynamo and functorch build suffice?

**Comment by csarofeen** [comment]
Yes. We don't do much special in our nightly builds and primarily rely on `setup.py`. So if there's logic in `setup.py` to pull working versions and build them under any flag, that would work great. @xwang233 in case there's any additional details of our CI/nightly process we should take into consideration.

**Comment by xwang233** [comment]
We currently build pytorch, torchvision, functorch, and torchdynamo directly from the latest master commits on github. I like the idea of `.github/ci_commit_pins/torchdynamo.txt` and we can try that.

**Comment by ezyang** [comment]
Yes please do. functorch doesn't have a pin yet; we could add something temporary unless @zou3519 thinks it isn't necessary

**Comment by zou3519** [comment]
To clarify, how does the pin work? My mental model so far has the following:
1. add a .github/ci_commit_pins/functorch.txt that has a commit hash of functorch
2. add a functorch build+test to pytorch/pytorch that does a build+test from the commit hash
3. any changes that break the functorch build+test... need an update in the functorch repo followed by a PR in the pytorch repo that updates the commit hash in the pin

(3) is one of the things I'm not sure about, because we also have an internal version of functorch that also needs updating. But this sounds no worse than the current situation

**Comment by malfet** [comment]
Yes, I plan to put a a PR that implements something like 1+2 today

**Comment by mruberry** [comment]
ping @malfet -- any updates here?

**Comment by malfet** [comment]
> ping @malfet -- any updates here?

Functorch is part of the repo, but dynamo remains a pinned dependency for the time being, but we need to continue discussion about what to do about it in the future

**Comment by xwang233** [comment]
We're now using the torchdynamo pinned commit from github pytorch repo in our internal CI.
